assets/src/edit-story/components/library/panes/shared/libraryMoveable.js

```diff
@@ -0,0 +1,228 @@
+  }));
+
+  const {
+    state: { activeDropTargetId },
+    actions: { setDraggingResource },
+  } = useDropTargets();
+
+  const frame = {
+    translate: [0, 0],
+  };
+
+  const eventTracker = useRef({});
+  const startEventTracking = (evt) => {
+    const { timeStamp, clientX, clientY } = evt;
+    eventTracker.current = {
+      timeStamp,
+      clientX,
+      clientY,
+    };
+  };
+
+  const onDrag = ({ beforeTranslate, inputEvent }) => {
+    frame.translate = beforeTranslate;
+    if (cloneRef.current) {
+      if (!cloneRef.current.style.opacity) {
+        // We're not doing it in `onDragStart` since otherwise on clicking it would appear, too.
+        cloneRef.current.style.opacity = 1;
+      }
+      cloneRef.current.style.transform = `translate(${beforeTranslate[0]}px, ${beforeTranslate[1]}px)`;
+      // We also have to move the original target ref for snapping to work.
+      targetBoxRef.current.style.transform = `translate(${beforeTranslate[0]}px, ${beforeTranslate[1]}px)`;
+    }
+    handleDrag(inputEvent);
+  };
+
+  const getTargetOffset = useCallback(() => {
+    const overlay = overlayRef.current;
+    let offsetX = 0,
+      offsetY = 0;
+    for (
+      let offsetNode = overlay;
+      offsetNode;
+      offsetNode = offsetNode.offsetParent
+    ) {
+      offsetX += offsetNode.offsetLeft;
+      offsetY += offsetNode.offsetTop;
+    }
+    return {
+      offsetX,
+      offsetY,
+    };
+  }, [overlayRef]);
+
+  const onDragStart = ({ set, inputEvent }) => {
+    // Note: we can't set isDragging true here since a "click" is also considered dragStart.
+    set(frame.translate);
+    setIsDragging(true);
+    startEventTracking(inputEvent);
+
+    // Position the clone that's being dragged.
+    const { offsetX, offsetY } = getTargetOffset();
+    const mediaBox = targetBoxRef.current.getBoundingClientRect();
+    const x1 = mediaBox.left - offsetX;
+    const y1 = mediaBox.top - offsetY;
+    cloneRef.current.style.left = `${x1}px`;
+    cloneRef.current.style.top = `${y1}px`;
+  };
+
+  const onDragEnd = ({ inputEvent }) => {
+    if (isMouseUpAClick(inputEvent, eventTracker.current)) {
+      resetMoveable();
+      onClick();
+      return false;
+    }
+    // We only skip Moveable onDragEnd handling if there's an active drop target ID.
+    if (activeDropTargetId) {
+      handleDragEnd();
+      // Only continue if the clone is at least partially on the page.
+    } else if (!isTargetOutOfContainer(cloneRef.current, pageContainer)) {
+      const {
+        x,
+        y,
+        width: w,
+        height: h,
+      } = cloneRef.current.getBoundingClientRect();
+      const { x: pageX, y: pageY } = pageContainer.getBoundingClientRect();
+
+      insertElement(type, {
+        ...elementProps,
+        x: editorToDataX(x - pageX, pageSize.width),
+        y: editorToDataY(y - pageY, pageSize.height),
+        width: editorToDataX(w, pageSize.width),
+        height: editorToDataY(h, pageSize.height),
```

Comment:
If you get these from the `UnitsProvider` via `useUnits()`, you don't need the page size argument - it's implicit. If it's possible to use that context of course?

Reply:
`useUnits()` is not available here, it seemed more straightforward to just use the available page size instead of reorganizing the structure for making `useUnits` available.